File: fridgebridge/lib/models/recipe.dart
class RecipeIngredient {
  final String ingredientId;
  final String name;
  final double quantity;
  final String unit;

  RecipeIngredient({
    required this.ingredientId,
    required this.name,
    required this.quantity,
    required this.unit,
  });

  factory RecipeIngredient.fromJson(Map<String, dynamic> json) =>
      RecipeIngredient(
        ingredientId: json['ingredientId'] ?? '',
        name: json['name'] ?? '',
        quantity: (json['quantity'] ?? 0).toDouble(),
        unit: json['unit'] ?? '',
      );

  Map<String, dynamic> toJson() => {
        'ingredientId': ingredientId,
        'name': name,
        'quantity': quantity,
        'unit': unit,
      };
}

class MissingIngredient {
  final String name;
  final double requiredQty;
  final double available;
  final String unit;

  MissingIngredient({
    required this.name,
    required this.requiredQty,
    required this.available,
    required this.unit,
  });

  factory MissingIngredient.fromJson(Map<String, dynamic> json) =>
      MissingIngredient(
        name: json['name'] ?? '',
        requiredQty: (json['required'] ?? 0).toDouble(),
        available: (json['available'] ?? 0).toDouble(),
        unit: json['unit'] ?? '',
      );
}

class Recipe {
  final String id;
  final String name;
  final String description;
  final List<RecipeIngredient> ingredients;
  final String availability;
  final List<MissingIngredient> missingIngredients;

  Recipe({
    required this.id,
    required this.name,
    required this.description,
    required this.ingredients,
    this.availability = 'unavailable',
    this.missingIngredients = const [],
  });

  factory Recipe.fromJson(Map<String, dynamic> json) => Recipe(
        id: json['id'] ?? '',
        name: json['name'] ?? '',
        description: json['description'] ?? '',
        ingredients: (json['ingredients'] as List<dynamic>?)
                ?.map((e) =>
                    RecipeIngredient.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
        availability: json['availability'] ?? 'unavailable',
        missingIngredients: (json['missing_ingredients'] as List<dynamic>?)
                ?.map((e) =>
                    MissingIngredient.fromJson(e as Map<String, dynamic>))
                .toList() ??
            [],
      );
}

File: backend/models/recipe.py
from pydantic import BaseModel, Field, field_validator
from typing import Optional, List

"""
Pydantic enforces type safety - this prevents users from being able to sneak in different data
types to cause damage to the database. Field constraints add length limits and value bounds
to prevent oversized payloads and negative quantities. field_validators strip and validate strings.
"""

class RecipeIngredient(BaseModel):
    ingredientId: str = Field(min_length=1, max_length=100)
    name: str = Field(min_length=1, max_length=100)
    quantity: float = Field(gt=0)
    unit: str = Field(min_length=1, max_length=20)

    @field_validator("name", "unit")
    @classmethod
    def strip_strings(cls, v: str) -> str:
        stripped = v.strip()
        if not stripped:
            raise ValueError("field cannot be blank")
        return stripped


class RecipeCreate(BaseModel):
    name: str = Field(min_length=1, max_length=100)
    description: str = Field(min_length=0, max_length=1000)
    ingredients: List[RecipeIngredient] = Field(min_length=1, max_length=50)

    @field_validator("name", "description")
    @classmethod
    def strip_strings(cls, v: str) -> str:
        return v.strip()


class RecipeUpdate(BaseModel):
    name: Optional[str] = Field(default=None, max_length=100)
    description: Optional[str] = Field(default=None, max_length=1000)
    ingredients: Optional[List[RecipeIngredient]] = Field(default=None, max_length=50)

    @field_validator("name", "description")
    @classmethod
    def strip_strings(cls, v: Optional[str]) -> Optional[str]:
        return v.strip() if v is not None else v


class MissingIngredient(BaseModel):
    name: str
    required: float
    available: float
    unit: str


class RecipeResponse(BaseModel):
    id: str
    name: str
    description: str
    ingredients: List[RecipeIngredient]
    availability: str = "unavailable"
    missing_ingredients: List[MissingIngredient] = []
